pkg/crc/preflight/preflight_check_tray_darwin.go

```diff
@@ -127,48 +105,111 @@ func fixTrayExistsAndRunning() error {
 		StdOutFilePath: stdOutFilePathDaemon,
 	}
 
-	trayConfig := AgentConfig{
-		BinaryPath:     filepath.Join(constants.CrcBinDir, constants.TrayBinaryName, "Contents", "MacOS", "CodeReady Containers"),
-		StdOutFilePath: stdOutFilePathTray,
-	}
 	logging.Debug("Creating daemon plist")
 	err = createPlist(daemonPlistTemplate, daemonConfig, daemonPlistFilePath)
 	if err != nil {
 		return err
 	}
+	// load plist and restart agent
+	if err := launchctlLoadPlist(daemonPlistFilePath); err != nil {
+		return err
+	}
+
+	if err := restartAgent(daemonAgentLabel); err != nil {
+		return err
+	}
+	return nil
+}
+
+func checkIfTrayPlistFileExists() error {
+	if !os.FileExists(trayPlistFilePath) {
+		return fmt.Errorf("Tray plist file does not exist")
+	}
+	return nil
+}
+
+func fixTrayPlistFileExists() error {
+	if err := ensureLaunchAgentsDirExists(); err != nil {
```

Comment:
This is very similar to `fixDaemonPlistFileExists`, for the longer methods in this file, it would be good to try to factor the common code in helper functions.

Reply:
There's really not much common code here, and already using the `createPlist` helper.. we can have an helper to generate the struct for the plist creation..

Reply:
```
diff --git a/pkg/crc/preflight/preflight_check_tray_darwin.go b/pkg/crc/preflight/preflight_check_tray_darwin.go
index c91e232f..4e727cc0 100644
--- a/pkg/crc/preflight/preflight_check_tray_darwin.go
+++ b/pkg/crc/preflight/preflight_check_tray_darwin.go
@@ -45,9 +45,6 @@ func checkIfDaemonPlistFileExists() error {
 }
 
 func fixDaemonPlistFileExists() error {
-	if err := ensureLaunchAgentsDirExists(); err != nil {
-		return err
-	}
 	currentExecutablePath, err := goos.Executable()
 	if err != nil {
 		return err
@@ -59,23 +56,7 @@ func fixDaemonPlistFileExists() error {
 		Args:           []string{"daemon", "--log-level", "debug"},
 	}
 
-	logging.Debug("Creating daemon plist")
-	err = launchd.CreatePlist(daemonConfig, daemonPlistFilePath)
-	if err != nil {
-		logging.Debug("failed while creating plist:", err.Error())
-		return err
-	}
-	// load plist and restart agent
-	if err := launchd.LoadPlist(daemonPlistFilePath); err != nil {
-		logging.Debug("Failed while loading plist:", err.Error())
-		return err
-	}
-
-	if err := launchd.RestartAgent(daemonAgentLabel); err != nil {
-		logging.Debug("Failed while restarting daemon:", err.Error())
-		return err
-	}
-	return nil
+	return fixPlistFileExists(daemonConfig, daemonPlistFilePath)
 }
 
 func checkIfTrayPlistFileExists() error {
@@ -85,29 +66,35 @@ func checkIfTrayPlistFileExists() error {
 	return nil
 }
 
+func fixPlistFileExists(agentConfig launchd.AgentConfig, plistFilePath string) error {
+	if err := ensureLaunchAgentsDirExists(); err != nil {
+		return err
+	}
+	logging.Debugf("Creating plist for %s", agentConfig.Label)
+	err := launchd.CreatePlist(agentConfig, plistFilePath)
+	if err != nil {
+		return err
+	}
+	// load plist and restart agent
+	if err := launchd.LoadPlist(plistFilePath); err != nil {
+		logging.Debug("failed while creating plist:", err.Error())
+		return err
+	}
+
+	if err := launchd.RestartAgent(agentConfig.Label); err != nil {
+		logging.Debug("Failed while loading plist:", err.Error())
+		return err
+	}
+	return nil
+}
 func fixTrayPlistFileExists() error {
-	if err := ensureLaunchAgentsDirExists(); err != nil {
-		return err
-	}
 	trayConfig := launchd.AgentConfig{
 		Label:          trayAgentLabel,
 		BinaryPath:     filepath.Join(constants.CrcBinDir, constants.TrayBinaryName, "Contents", "MacOS", "CodeReady Containers"),
 		StdOutFilePath: stdOutFilePathTray,
 	}
-	logging.Debug("Creating tray plist")
-	err := launchd.CreatePlist(trayConfig, trayPlistFilePath)
-	if err != nil {
-		return err
-	}
-	// load plist and restart agent
-	if err := launchd.LoadPlist(trayPlistFilePath); err != nil {
-		return err
-	}
 
-	if err := launchd.RestartAgent(trayAgentLabel); err != nil {
-		return err
-	}
-	return nil
+	return fixPlistFileExists(trayConfig, trayPlistFilePath)
 }
 
 func checkIfDaemonAgentRunning() error {
```